### О задании

Задание состоит из двух разделов, посвященных работе с табличными данными с помощью библиотеки pandas и визуализации с помощью matplotlib. В первом разделе вам предстоит выполнить базовые задания с помощью вышеуказанных библиотек, а во втором распределить студентов по курсам. Баллы даются за выполнение отдельных пунктов. Задачи в рамках одного раздела рекомендуется решать в том порядке, в котором они даны в задании.

Задание направлено на освоение jupyter notebook (будет использоваться в дальнейших заданиях), библиотек pandas и matplotlib.

### Оценивание и штрафы
Каждая из задач имеет определенную «стоимость» (указана в скобках около задачи). Максимально допустимая оценка за работу — 10 баллов.

Сдавать задание после жёсткого дедлайна нельзя. При выставлении неполного балла за задание в связи с наличием ошибок на усмотрение проверяющего предусмотрена возможность исправить работу на указанных в ответном письме условиях.

Задание выполняется самостоятельно. «Похожие» решения считаются плагиатом и все задействованные студенты (в том числе те, у кого списали) не могут получить за него больше 0 баллов (подробнее о плагиате см. на странице курса). Если вы нашли решение какого-то из заданий (или его часть) в открытом источнике, необходимо указать ссылку на этот источник (скорее всего вы будете не единственным, кто это нашел, поэтому чтобы исключить подозрение в плагиате, необходима ссылка на источник).

## 0. Введение

Сейчас мы находимся в jupyter-ноутбуке (или ipython-ноутбуке). Это удобная среда для написания кода, проведения экспериментов, изучения данных, построения визуализаций и других нужд, не связанных с написанием production-кода. 

Ноутбук состоит из ячеек, каждая из которых может быть либо ячейкой с кодом, либо ячейкой с текстом размеченным и неразмеченным. Текст поддерживает markdown-разметку и формулы в Latex.

Для работы с содержимым ячейки используется *режим редактирования* (*Edit mode*, включается нажатием клавиши **Enter** после выбора ячейки), а для навигации между ячейками искользуется *командный режим* (*Command mode*, включается нажатием клавиши **Esc**). Тип ячейки можно задать в командном режиме либо с помощью горячих клавиш (**y** to code, **m** to markdown, **r** to edit raw text), либо в меню *Cell -> Cell type*. 

После заполнения ячейки нужно нажать *Shift + Enter*, эта команда обработает содержимое ячейки: проинтерпретирует код или сверстает размеченный текст.

In [128]:
# ячейка с кодом, при выполнении которой появится output
2 + 2

4

Ячейка с неразмеченным текстом.

Попробуйте создать свои ячейки, написать какой-нибудь код и текст какой-нибудь формулой.

In [129]:
4 * 10

40

[Здесь](https://athena.brynmawr.edu/jupyter/hub/dblank/public/Jupyter%20Notebook%20Users%20Manual.ipynb) находится <s>не</s>большая заметка о используемом языке разметки Markdown. Он позволяет:

0. Составлять упорядоченные списки
1. Выделять *текст* <s>при</s> **необходимости**
2. Добавлять [ссылки](http://imgs.xkcd.com/comics/the_universal_label.png)


* Составлять неупорядоченные списки

Делать вставки с помощью LaTex:
    
$
\left\{
\begin{array}{ll}
x = 16 \sin^3 (t) \\ 
y = 13 \cos (t) - 5 \cos (2t) - 2 \cos (3t) - \cos (4t) \\
t \in [0, 2 \pi]
\end{array}
\right.$

А ещё можно вставлять картинки, или гифки, или что захотите:

<img src="https://media1.tenor.com/images/9b8fbe9214504bcf6c60fe4e4f7e114e/tenor.gif?itemid=5416416" style="width: 400px">

### Google Colab

Что за колаб? 

**Google Colab (Colaboratory)** это **Jupyter Notebook + Cloud + Google Drive.**

Компания Google предоставляет возможность бесплатно запускать ноутбуки (предварительно загрузив их на свой гугл-диск) прямо в облаке. При этом вам не требуется установка никаких пакетов на свою машину, а работать можно прямиком из браузера. Вот ссылка:

https://colab.research.google.com

При использовании вы увидете много сходств с jupyter ноутбуком. Одним из преимуществ является доступность GPU, соответствующую опцию можно активировать в настройках сервиса. При желании вы сможете найти в интернете много туториалов по использованию или разобраться самостоятельно =)

## 1. Табличные данные и Pandas

Pandas — удобная библиотека для работы с табличными данными в Python, если данных не слишком много и они помещаются в оперативную память вашего компьютера. Несмотря на неэффективность реализации и некоторые проблемы, библиотека стала стандартом в анализе данных.

Основной объект в pandas — это DataFrame, представляющий собой таблицу с именованными колонками различных типов, индексом (может быть многоуровневым). DataFrame можно создавать, считывая таблицу из файла или задавая вручную из других объектов.

В этой части потребуется выполнить несколько небольших заданий. Можно пойти двумя путями: сначала изучить материалы, а потом приступить к заданиям, или же разбираться "по ходу". Выбирайте сами.

Материалы:
1. [Pandas за 10 минут из официального руководства](http://pandas.pydata.org/pandas-docs/stable/10min.html)
2. [Документация](http://pandas.pydata.org/pandas-docs/stable/index.html) (стоит обращаться, если не понятно, как вызывать конкретный метод)
3. [Примеры использования функционала](http://nbviewer.jupyter.org/github/justmarkham/pandas-videos/blob/master/pandas.ipynb)

Многие из заданий можно выполнить несколькими способами. Не существуют единственно верного, но попробуйте максимально задействовать арсенал pandas и ориентируйтесь на простоту и понятность вашего кода. Мы не будем подсказывать, что нужно использовать для решения конкретной задачи, попробуйте находить необходимый функционал сами (название метода чаще всего очевидно). В помощь вам документация, поиск и stackoverflow.

In [71]:
import pandas as pd

<span style="color:red">Перед выполнением задания необходимо ознакомиться с первым семинаром (seminars/sem01-pandas.ipynb) </span>.

Это поможет вам получить общее понимание происходяшего и успешнее справиться с заданием.

Также успешному выполнению способствует внимательное чтение текста задания от начала до конца.

Удачи!

Скачаем подготовленные на семинаре данные:

In [1]:
# !wget  -O 'end_seminar.xlsx' -q 'https://www.dropbox.com/s/f4rm8sjc3v99p0m/_end_seminar.xlsx?dl=0'

zsh:1: command not found: wget


Для пользователей Windows: скачайте файл самостоятельно и поместите его в папку с тетрадкой. Или попробуйте один из следующих вариантов:

In [ ]:
# !powershell iwr -outf somefile https://somesite/somefile

In [246]:
# !pip install wget
import wget 
wget.download('https://dropbox.com/s/f4rm8sjc3v99p0m/_end_seminar.xlsx?dl=1', 'end_seminar.xlsx')

100% [..........................................................] 50735 / 50735

'end_seminar (1).xlsx'

##### В первой части задания (до раздела "Распределение студентов по курсам") использование циклов запрещается и повлечет за собой снижение оценки. Использование <code>vectorize</code> и  <code>apply</code>, <code>apply_along_axis</code> крайне нежелательно.

Для каждой задачи из этого раздела вы должны написать код для получения ответа, а также дать текстовый ответ, если он предполагается.

На некоторые вопросы вы можете получить путём пристального взгляда на таблицу, но это не будет засчитываться. Вы в любом случае должны получить ответ с помощью кода.

#### 1. [0 баллов] Откройте файл с таблицей (не забудьте про её формат). Выведите последние 10 строк.

Посмотрите на данные и скажите, что они из себя представляют, сколько в таблице строк, какие столбцы? (на это не надо отвечать, просто подумайте об этом)

In [247]:
import pandas as pd
df = pd.read_excel("end_seminar.xlsx")
df.tail(10)

timestamp                                id  rating  \
351 2020-05-18 23:58:33.903  8fe79f84f36e3a5d2d6745621321302c    1016   
352 2020-05-18 23:59:20.831  f93afd8e92c336b471aa3bbbc579aefe    1298   
353 2020-05-18 23:59:28.998  9963fc00d9405a35b894cc04a7e19855    1135   
354 2020-05-19 08:55:39.509  98c5a2548d93f26bd4195466e93178b3    1035   
355 2020-05-19 18:58:06.283  ba298cbd9be59b84bab33325e956f27e    1155   
356 2020-05-19 19:04:17.873  c5870b37ee3f672e4259dfb2fc02ed8f    1029   
357 2020-05-19 20:19:25.815  f44aea1095f9f5b8e2f5f0ed2b17891b    1060   
358 2020-05-20 11:54:38.318  d7118b44c7e8fdbbaaca226d2da6fef1    1111   
359 2020-05-23 21:16:40.916  79d14db48e4737af731960ae212d5134     966   
360 2020-05-23 21:16:40.916  f2d210ace25a841e1e58036b257d5d47    1195   

     18_group  is_mi                                           fall_1  \
351       NaN    NaN                 Безопасность компьютерных систем   
352       NaN    NaN                Высокопроизводительные вычисления   
353       NaN    NaN                 Безопасность компьютерных систем   
354       NaN    NaN            Байесовские методы машинного обучения   
355       NaN    NaN  Сбор и обработка данных с помощью краудсорсинга   
356       NaN    NaN                      Statistical Learning Theory   
357       NaN    NaN                Анализ неструктурированных данных   
358       NaN    NaN  Сбор и обработка данных с помощью краудсорсинга   
359       NaN    NaN                 Безопасность компьютерных систем   
360       NaN    NaN                Анализ неструктурированных данных   

                                                fall_2  \
351                  Высокопроизводительные вычисления   
352    Сбор и обработка данных с помощью краудсорсинга   
353                  Высокопроизводительные вычисления   
354               Глубинное обучение в обработке звука   
355              Байесовские методы машинного обучения   
356                  Анализ неструктурированных данных   
357                  Высокопроизводительные вычисления   
358                        Statistical Learning Theory   
359  Принятие решений в условиях риска и неопределё...   
360                      Моделирование временных рядов   

                                              fall_3  \
351  Сбор и обработка данных с помощью краудсорсинга   
352                      Statistical Learning Theory   
353                 Безопасность компьютерных систем   
354                      Statistical Learning Theory   
355                Анализ неструктурированных данных   
356  Сбор и обработка данных с помощью краудсорсинга   
357                    Моделирование временных рядов   
358                Анализ неструктурированных данных   
359                Анализ неструктурированных данных   
360                Высокопроизводительные вычисления   

                                              spring_1  \
351                             Конфликты и кооперация   
352               Методы сжатия и передачи медиаданных   
353  Проектирование и разработка высоконагруженных ...   
354                           Обучение с подкреплением   
355                             Конфликты и кооперация   
356                           Обучение с подкреплением   
357                             Конфликты и кооперация   
358  Проектирование и разработка высоконагруженных ...   
359  Проектирование и разработка высоконагруженных ...   
360                           Обучение с подкреплением   

                                              spring_2  \
351  Проектирование и разработка высоконагруженных ...   
352  Проектирование и разработка высоконагруженных ...   
353               Методы сжатия и передачи медиаданных   
354            Генеративные модели в машинном обучении   
355                           Обучение с подкреплением   
356                             Конфликты и кооперация   
357                           Обучение с подкреплением   
358               Методы сжатия и передачи медиадан

In [93]:
df.head()

timestamp                                id  rating  18_group  \
0 2020-05-15 01:12:50.543  93ff79a51cd602f1dd3028ba2c129503     698     181.0   
1 2020-05-15 02:46:48.066  26b01b1c4cd5656bab18d24c548834fb     647     181.0   
2 2020-05-15 03:12:41.480  30f3653fc176d54e89ac3179c455c6dd     624     185.0   
3 2020-05-15 04:43:08.994  1528f0eaa027580820ccf0d92a53ad68     579     182.0   
4 2020-05-15 07:47:17.197  496ea4f0d4abe264b1bb1b80eb3830c5     572     183.0   

   is_mi                             fall_1  \
0    NaN                           Язык SQL   
1    NaN  Высокопроизводительные вычисления   
2    NaN   Безопасность компьютерных систем   
3    NaN        Statistical Learning Theory   
4    NaN  Высокопроизводительные вычисления   

                              fall_2                         fall_3  \
0  Высокопроизводительные вычисления           Матричные вычисления   
1   Безопасность компьютерных систем                       Язык SQL   
2               Матричные вычисления  Моделирование временных рядов   
3  Высокопроизводительные вычисления           Матричные вычисления   
4   Безопасность компьютерных систем              Теория баз данных   

                 spring_1                spring_2             spring_3  \
0  Дискретная оптимизация        Численные методы  Машинное обучение 2   
1  Дискретная оптимизация        Численные методы  Машинное обучение 2   
2  Дискретная оптимизация     Машинное обучение 2     Численные методы   
3  Дискретная оптимизация        Численные методы  Машинное обучение 2   
4       Компьютерные сети  Дискретная оптимизация     Численные методы   

  is_first_time  percentile  17_group blended  is_ml_student  
0            Да    0.183857       NaN     NaN            NaN  
1           Нет    0.426009       NaN     NaN            NaN  
2            Да    0.538117       NaN     NaN            NaN  
3            Да    0.744395       NaN     NaN            NaN  
4            Да    0.780269       NaN     NaN            NaN

#### 2. [0.5 балла] Есть ли в данных пропуски? В каких колонках? Сколько их в каждой из этих колонок?

In [94]:
a = df.isnull()
check_nan = a.values.any()
print(a.values.any())
print(a.sum())

True
timestamp          0
id                 0
rating             0
18_group         138
is_mi            343
fall_1             0
fall_2             0
fall_3             0
spring_1           0
spring_2           0
spring_3           0
is_first_time      2
percentile         0
17_group         223
blended          223
is_ml_student    304
dtype: int64


Заполните пропуски пустой строкой для строковых колонок и нулём для числовых (постарайтесь избежать перечисления названий всех столбцов).

In [96]:
print(df['18_group'].dtypes, df['17_group'].dtypes, df['blended'].dtypes, df['is_mi'].dtypes, df['is_first_time'].dtypes, df['is_ml_student'].dtypes)

df['18_group'] = df['18_group'].fillna(0)
df['17_group'] = df['17_group'].fillna(0)
df['blended'] = df['blended'].fillna('')
df['is_mi'] = df['is_mi'].fillna(0)
df['is_first_time'] = df['is_first_time'].fillna('')
df['is_ml_student'] = df['is_ml_student'].fillna(0)

float64 float64 object float64 object float64


In [97]:
df.tail()

timestamp                                id  rating  \
356 2020-05-19 19:04:17.873  c5870b37ee3f672e4259dfb2fc02ed8f    1029   
357 2020-05-19 20:19:25.815  f44aea1095f9f5b8e2f5f0ed2b17891b    1060   
358 2020-05-20 11:54:38.318  d7118b44c7e8fdbbaaca226d2da6fef1    1111   
359 2020-05-23 21:16:40.916  79d14db48e4737af731960ae212d5134     966   
360 2020-05-23 21:16:40.916  f2d210ace25a841e1e58036b257d5d47    1195   

     18_group  is_mi                                           fall_1  \
356       0.0    0.0                      Statistical Learning Theory   
357       0.0    0.0                Анализ неструктурированных данных   
358       0.0    0.0  Сбор и обработка данных с помощью краудсорсинга   
359       0.0    0.0                 Безопасность компьютерных систем   
360       0.0    0.0                Анализ неструктурированных данных   

                                                fall_2  \
356                  Анализ неструктурированных данных   
357                  Высокопроизводительные вычисления   
358                        Statistical Learning Theory   
359  Принятие решений в условиях риска и неопределё...   
360                      Моделирование временных рядов   

                                              fall_3  \
356  Сбор и обработка данных с помощью краудсорсинга   
357                    Моделирование временных рядов   
358                Анализ неструктурированных данных   
359                Анализ неструктурированных данных   
360                Высокопроизводительные вычисления   

                                              spring_1  \
356                           Обучение с подкреплением   
357                             Конфликты и кооперация   
358  Проектирование и разработка высоконагруженных ...   
359  Проектирование и разработка высоконагруженных ...   
360                           Обучение с подкреплением   

                                 spring_2  \
356                Конфликты и кооперация   
357              Обучение с подкреплением   
358  Методы сжатия и передачи медиаданных   
359  Методы сжатия и передачи медиаданных   
360                   Компьютерное зрение   

                                 spring_3 is_first_time  percentile  17_group  \
356                   Компьютерное зрение            Да    0.920290     174.0   
357                      Численные методы            Да    0.840580     174.0   
358                      Численные методы            Да    0.652174     173.0   
359                   Компьютерное зрение            Да    0.985507     172.0   
360  Методы сжатия и передачи медиаданных                  0.369565     176.0   

                            blended  is_ml_student  
356                          DevOps            0.0  
357                          DevOps            0.0  
358  Соревновательный анализ данных            0.0  
359                          DevOps            0.0  
360  Соревновательный анализ данных            0.0

#### 3. [0.5 балла] Посмотрите повнимательнее на колонку 'is_first_time'. 

Есть ли в ней ответы "Нет"? Сколько их?

Если вы найдете повторные обращения студентов, оставьте только самую позднюю версию. В дальнейших заданиях используйте версию данных без повторов.

<i>Обращения со значением "Нет" в 'is_first_time' могут быть как повторными, так и первичными, поскольку поле заполняли сами студенты.</i>

In [248]:
df.drop_duplicates(subset = 'id', keep = 'last', inplace=True, ignore_index=True)
print(df['is_first_time'].value_counts())
#https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.drop_duplicates.html

Да     294
Нет     51
Name: is_first_time, dtype: int64


#### 4. [0.5 балла] Ответьте на вопросы:
1. Сколько было заявок из групп 18-го года набора, а сколько из групп 17-го года?
2. Есть ли студенты с равными перцентилями (среди объединенных данных, ведь конкурс на каждый курс общий)?

In [249]:
print((df.isnull()).values.any())
df.info()

True
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 347 entries, 0 to 346
Data columns (total 16 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   timestamp      347 non-null    datetime64[ns]
 1   id             347 non-null    object        
 2   rating         347 non-null    int64         
 3   18_group       213 non-null    float64       
 4   is_mi          16 non-null     float64       
 5   fall_1         347 non-null    object        
 6   fall_2         347 non-null    object        
 7   fall_3         347 non-null    object        
 8   spring_1       347 non-null    object        
 9   spring_2       347 non-null    object        
 10  spring_3       347 non-null    object        
 11  is_first_time  345 non-null    object        
 12  percentile     347 non-null    float64       
 13  17_group       134 non-null    float64       
 14  blended        134 non-null    object        
 15  is_ml_student  56 

In [250]:
#1.
print("18 : ", df['18_group'].astype(bool).sum(axis=0))
print("17 : ", df['17_group'].astype(bool).sum(axis=0))
#https://stackoverflow.com/questions/26053849/counting-non-zero-values-in-each-column-of-a-dataframe-in-python

18 :  347
17 :  347


In [112]:
#2. https://stackoverflow.com/questions/50242968/check-for-duplicate-values-in-pandas-dataframe-column
print(len(df['percentile'].unique()) < len(df.index))

True


#### 5. [0.5 балла] Какие  blended-курсы для четверокурсников существуют? На какой blended-курс записалось наибольшее количество студентов? На каком из blended-курсов собрались студенты с самым высоким средним рейтингом (выведите этот курс и количество студентов на нем)? 

In [130]:
print(df['blended'].value_counts())
print(df.groupby('blended')['rating'].apply(lambda x : x.mean()), end = '\n\n')
print((df['blended'] == 'Введение в дифференциальную геометрию').sum())
#https://www.width.ai/pandas/count-specific-value-in-column-with-pandas

                                         213
DevOps                                    87
Соревновательный анализ данных            32
Введение в дифференциальную геометрию     15
Name: blended, dtype: int64
blended
                                          633.901408
DevOps                                   1154.172414
Введение в дифференциальную геометрию    1210.733333
Соревновательный анализ данных           1144.218750
Name: rating, dtype: float64

15


#### 6. [1 балл] Выясните, есть ли студенты с абсолютно одинаковыми предпочтениями по всем курсам.

Для этого сформируйте таблицу, где для каждого возможного набора курсов посчитано количество студентов, выбравших такой набор, и оставьте только строки где это количество больше 1.

В данном случае набор курсов задается упорядоченным множеством ('fall_1', 'fall_2', 'fall_3', 'spring_1', 'spring_2', 'spring_3', 'blended'). Элемент blended будет нулевым для 3-го курса.

In [174]:
#https://pbpython.com/pandas-pivot-table-explained.html
table = pd.pivot_table(df, index=['fall_1', 'fall_2', 'fall_3', 'spring_1', 'spring_2', 'spring_3', 'blended'], values='id', aggfunc=(lambda x: len(x) if len(x)>1 else pd.NaT))
#table.loc[1]
table.dropna()

id
fall_1                                          fall_2                                          fall_3                                          spring_1                                           spring_2                                    spring_3                                    blended   
Statistical Learning Theory                     Анализ неструктурированных данных               Сбор и обработка данных с помощью краудсорсинга Обучение с подкреплением                           Конфликты и кооперация                      Компьютерное зрение                         DevOps   2
Байесовские методы машинного обучения           Анализ неструктурированных данных               Глубинное обучение в обработке звука            Обучение с подкреплением                           Генеративные модели в машинном обучении     Компьютерное зрение                         DevOps   2
                                                                                                Сбор и обработка данных с помощью краудсорсинга Обучение с подкреплением                           Компьютерное зрение                         Генеративные модели в машинном обучении     DevOps   2
Безопасность компьютерных систем                Statistical Learning Theory                     Матричные вычисления                            Дискретная оптимизация                             Промышленное программирование на языке Java Дополнительные главы прикладной статистики           3
                                                Высокопроизводительные вычисления               Анализ неструктурированных данных               Проектирование и разработка высоконагруженных с... Методы сжатия и передачи медиаданных        Конфликты и кооперация                      DevOps   2
                                                                                                Язык SQL                                        Компьютерные сети                                  Промышленное программирование на языке Java Дискретная оптимизация                               3
                                                                                                                                                Промышленное программирование на языке Java        Компьютерные сети                           Системы баз данных                                   2
                                                Матричные вычисления                            Моделирование временных рядов                   Промышленное программирование на языке Java        Компьютерные сети                           Дискретная оптимизация                               2
                                                Сбор и обработка данных с помощью краудсорсинга Матричные вычисления                            Компьютерные сети                                  Промышленное программирование на языке Java Анализ данных в бизнесе                              2
                                                Теория баз данных                               Язык SQL                                        Машинное обучение 2                                Компьютерные сети                           Промышленное программирование на языке Java          2
Высокопроизводительные вычисления               Безопасность компьютерных систем                Сбор и обработка данных с помощью краудсорсинга Проектирование и разработка высоконагруженных с... Методы сжатия и передачи медиаданных        Компьютерное зрение                         DevOps   2
                                                Сбор и обработка данных с помощью краудсорсинга Моделирование временных рядов                   Проектирование и разработка высоконагруженных с... Методы сжатия и передачи медиаданных        Методы сжатия и передачи медиаданных        DevOps   2
Моделирование временных рядов                   Сбор и обработка данных с помощью краудсорсинга Безопасность компьютерных систем                Компьютерные сети              

#### 7. [0.5 балла] Найдите курсы по выбору, на которые записывались как студенты 18-го года набора, так и студенты 17-го года.

In [198]:
s1 = set(df[df.where('18_group' !=0)]['blended'])

ValueError: Array conditional must be same shape as self

Методом исключения найдите курсы, которые предлагались только студентам 18-го года и только студентам 17-го года.

In [ ]:
# your code

### Визуализации и matplotlib

При работе с данными часто неудобно делать какие-то выводы, если смотреть на таблицу и числа в частности, поэтому важно уметь визуализировать данные. Здесь будут описаны ключевые правила оформления графиков для **всех** домашних заданий.

У matplotlib, конечно же, есть [документация](https://matplotlib.org/users/index.html) с большим количеством [примеров](https://matplotlib.org/examples/), но для начала достаточно знать про несколько основных типов графиков:
- plot — обычный поточечный график, которым можно изображать кривые или отдельные точки;
- hist — гистограмма, показывающая распределение некоторой величины;
- scatter — график, показывающий взаимосвязь двух величин;
- bar — столбцовый график, показывающий взаимосвязь количественной величины от категориальной.

В этом задании вы попробуете построить один из них. Не забывайте про базовые принципы построения приличных графиков:
- оси должны быть подписаны, причём не слишком мелко;
- у графика должно быть название;
- если изображено несколько графиков, то необходима поясняющая легенда;
- все линии на графиках должны быть чётко видны (нет похожих цветов или цветов, сливающихся с фоном);
- если отображена величина, имеющая очевидный диапазон значений (например, проценты могут быть от 0 до 100), то желательно масштабировать ось на весь диапазон значений (исключением является случай, когда вам необходимо показать малое отличие, которое незаметно в таких масштабах);
- сетка на графике помогает оценить значения в точках на глаз, это обычно полезно, поэтому лучше ее отрисовывать;
- если распределение на гистограмме имеет тяжёлые хвосты, лучше использовать логарифмическую шкалу.

Еще одна билиотека для визуализации: [seaborn](https://jakevdp.github.io/PythonDataScienceHandbook/04.14-visualization-with-seaborn.html) (обычно сокращается до sns). Это настройка над matplotlib, иногда удобнее и красивее делать визуализации через неё. 

**5 пунктов после которых ваши графики не будут прежними:**
- для красивой картинки <code>%config InlineBackend.figure_format = 'retina'</code>
- задаем размер (почти) всех графиков (можно выбрать другие значения) <code>plt.rcParams['figure.figsize'] = 8, 5</code>
- размер шрифта подписей графиков <code>plt.rcParams['font.size'] = 12</code>
- формат в котором сохраняется изображение <code>mpl.rcParams['savefig.format'] = 'pdf'</code>
- sns – seaborn, добавляет решетку <code>sns.set_style('darkgrid')</code>

Добавьте эти функции в ячейку ниже. Каждый график все равно придется настраивать отдельно, но указанные строчки позволят значительно упростить процесс.  

In [181]:
%matplotlib inline
import matplotlib.pyplot as plt
import seaborn as sns

# your code

#### 8. [0.5 балла] Постройте график средних рейтингов по дням получения ответов (bar plot).

In [3]:
#https://stackoverflow.com/questions/48238305/bar-plot-with-groupby
df.groupby('timestamp')['rating'].apply(lambda x : x.mean()).plot()
plt.savefig("bar_graph.pdf", format="pdf")

NameError: name 'df' is not defined

Сохраните график в формате pdf (так он останется векторизованным).

In [252]:
# https://www.tutorialspoint.com/save-the-plots-into-a-pdf-in-matplotlib
plt.savefig("bar_graph.pdf", format="pdf")

<Figure size 640x480 with 0 Axes>

### 2. Распределение студентов по курсам.

<span style="color:red">!!!ВНИМАТЕЛЬНО ИЗУЧИТЕ ТЕКСТ НИЖЕ!!!</span>.

<span style="color:orange">Если во время выполнения заданий у вас вознинут вопросы -- еще раз перечитайте текст целиком, скорее всего ответы уже содержатся в нем.</span>

Теперь вам нужно распределить студентов по осенним курсам по выбору, учитывая их предпочтения.

Алгоритм распределения студентов по курсам:
1. По умолчанию на каждой дисциплине по выбору у 3 и 4 курсов может учиться 1 группа (до 30 студентов). Исключения описаны ниже. На blended-дисциплинах для четверокурсников количество мест не ограничено.
2. Проводится первая волна отбора. Для каждой дисциплины формируется список тех, кто указал её первым приоритетом (если студент должен выбрать два курса по выбору, то для него дисциплины, которые он указал первым и вторым приоритетом, рассматриваются как дисциплины первого приоритета). Если желающих больше, чем мест, то выбирается топ по перцентилю рейтинга.
3. На дисциплинах, где остались места после первой волны, формируются списки тех, кто выбрал их вторым приоритетом и еще не прошел на необходимое количество курсов. Места заполняются лучшими по перцентилю рейтинга студентами. После этого проводится такая же процедура для дисциплин третьего приоритета.
4. Если студент не попал на необходимое количество курсов по итогам трёх волн, с ним связывается учебный офис и решает вопрос в индивидуальном порядке.

Обращаем ваше внимание на следующие детали:
- Конкурс на каждый курс общий для 3-го и 4-го курса

- По умолчанию студент выбирает один осенний и один весенний курс по выбору, а также четверокурсники выбирают один blended-курс. Студенты групп 17-го года специализаций МОП и ТИ выбирают по 2 осенних и 2 весенних курса по выбору, также студенты групп 18' специализации МИ выбирают 2 осенних курса. <i>Для студентов, которые выбирают 2 курса (например, осенних) первый приоритет — <code>fall_1</code> и <code>fall_2</code>, второй приоритет — <code>fall_3</code>. Такие студенты участвуют только в двух волнах отбора</i>.

- Студенты специализации МОП не могут выбрать весенним курсом по выбору Машинное обучение 2. **Если студент специализации МОП выбрал Машинное обучение 2, то его приоритеты сдвигаются.** Из-за совпадений первого и второго курса по выбору двигать приоритеты не надо.

- Blended-курсы не трогайте, по ним не надо распределять, на другие курсы они никак не влияют.

- Заведомо известно, что в процессе распределения не возникнет ситуации, когда на одно место претендуют студенты с одинаковым перцентилем.

- Постарайтесь воздержаться от использования циклов там, где это возможно. <i>Допустимо итерироваться по <b>курсам</b>, на которые проводится отбор, и по <b>волнам</b> отбора. Если вы придумаете, как обойтись и без этих циклов, то на усмотрение проверяющего могут быть добавлены бонусные баллы. <b>Дублирование кода не признается успешным избавлением от циклов</b></i>

- На выходе ожидается файл res_fall.csv с результатами распределения на осенние курсы по выбору. Файл должен быть следующего формата:

    * Три колонки: ID, course1, course2
    
    * Если студент не попал на курс, но должен был, то вместо названия курса в ячейке должна быть строка "???"
    
    * Если студент должен выбрать только один курс, то в колонке course2 для него должна стоять строка "-"
    
    * Если студент должен выбрать два курса по выбору, то порядок в колонках course1 и course2 не важен.
    
    * hint: для сохранения воспользуйтесь df.to_csv('solution.csv', index=None)
    

Для работы вам могут понадобиться следующие данные:

- Результаты опроса (вы уже использовали этот файл в первой части задания, но на всякий случай ссылка: https://www.dropbox.com/s/f4rm8sjc3v99p0m/_end_seminar.xlsx?dl=0)

- Соответствие номеров групп специализациям:

    * 171, 172 - МОП; 173 - ТИ; 174 — АДИС; 175, 176 — РС; 177 — АПР
    
    * У студентов 18-го года номера групп соответствуют номерам до распределения по специализациям. Это означает, что по номеру группы 18* нельзя однозначно определить специализацию студента. При этом в рамках распределения важно знать информацию только о двух из них: МОП и МИ. Эти знание можно получить из колонок 'is_ml_student
' и 'is_mi' соответственно.

- Ограничения по количеству мест на курсах по выбору:

    * Осенние: везде 30 мест, кроме Statistical Learning Theory (60 мест), Высокопроизводительных вычислений (60 мест), Анализа неструктурированных данных ($\infty$ мест)

    * Весенние: везде 30 мест, кроме Обучения с подкреплением (60 мест), Анализа данных в бизнесе (60 мест).


Кстати, убедитесь, что в данных больше нет пропусков и повторных записей.

#### 0. Проверка

Для начала давайте убедимся, что вы успешно выполнили задания первой части и проверим ваши данные на наличие пропусков и повторов:

In [199]:
assert df.shape[0] == 347, 'В таблице остались повторы или потеряны данные'

assert df.isna().sum().sum() == 0, 'В таблице остались пропуски'

Если вы не получили AssertionError, то можете продолжать.

#### 1. [1 балл] Создайте новый признак, обозначающий, сколько осенних курсов должен выбрать студент

В этом вам может помочь информация о специализации и группе стундента.

In [254]:
# (￣^￣)ゞ
df['num'] = [2 if x == 171 or x == 172 else 1 for x in (df['17_group'])]
#df['num'] = [2 if x == 1.0 df[]
df['num'] = [2 if df.iloc[i, 4] == 1 else df.iloc[i, 16] for i in range(347)]
df.head()
#https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.iloc.html

timestamp                                id  rating  18_group  \
0 2020-05-15 01:12:50.543  93ff79a51cd602f1dd3028ba2c129503     698     181.0   
1 2020-05-15 02:46:48.066  26b01b1c4cd5656bab18d24c548834fb     647     181.0   
2 2020-05-15 03:12:41.480  30f3653fc176d54e89ac3179c455c6dd     624     185.0   
3 2020-05-15 04:43:08.994  1528f0eaa027580820ccf0d92a53ad68     579     182.0   
4 2020-05-15 07:47:17.197  496ea4f0d4abe264b1bb1b80eb3830c5     572     183.0   

   is_mi                             fall_1  \
0    NaN                           Язык SQL   
1    NaN  Высокопроизводительные вычисления   
2    NaN   Безопасность компьютерных систем   
3    NaN        Statistical Learning Theory   
4    NaN  Высокопроизводительные вычисления   

                              fall_2                         fall_3  \
0  Высокопроизводительные вычисления           Матричные вычисления   
1   Безопасность компьютерных систем                       Язык SQL   
2               Матричные вычисления  Моделирование временных рядов   
3  Высокопроизводительные вычисления           Матричные вычисления   
4   Безопасность компьютерных систем              Теория баз данных   

                 spring_1                spring_2             spring_3  \
0  Дискретная оптимизация        Численные методы  Машинное обучение 2   
1  Дискретная оптимизация        Численные методы  Машинное обучение 2   
2  Дискретная оптимизация     Машинное обучение 2     Численные методы   
3  Дискретная оптимизация        Численные методы  Машинное обучение 2   
4       Компьютерные сети  Дискретная оптимизация     Численные методы   

  is_first_time  percentile  17_group blended  is_ml_student  num  
0            Да    0.183857       NaN     NaN            NaN    1  
1           Нет    0.426009       NaN     NaN            NaN    1  
2            Да    0.538117       NaN     NaN            NaN    1  
3            Да    0.744395       NaN     NaN            NaN    1  
4            Да    0.780269       NaN     NaN            NaN    1

Проверка:

In [255]:
col_name = 'num'    # insert your new column name as str

assert(df[df['id'] == '2662600c2c37e11e62f6ee0b88452f22'][col_name] == 2).all()
assert(df[df['id'] == 'd555d2805e1d93d4f023e57dc4c8f403'][col_name] == 2).all()
assert(df[df['id'] == '8fe79f84f36e3a5d2d6745621321302c'][col_name] == 1).all()
assert(df[df['id'] == 'e4caca755ee0bdd711e18fb8084958b5'][col_name] == 1).all()

#### 2. [2 балла] Распределите студентов в соответствии с первым приоритетом

In [245]:
# (￣^￣)ゞ
import collections as cl 
m = cl.defaultdict(list)
for index, x in df.iterrows():
    if x['num'] == 2:
        m[x['fall_1']].append((x['rating'],x['id'], True))
        m[x['fall_2']].append((x['rating'],x['id'], True))
    else:
        m[x['fall_1']].append((x['rating'],x['id'], False))
for key, value in m.items():
    value.sort()
    if key == 'Statistical Learning Theory' or key == 'Высокопроизводительные вычисления':
        value = value[:60]
        continue
    if key == 'Анализ неструктурированных данных':
        continue
    value = value[:30]
    
def DictKey(a):
    return a[1]

ans = cl.defaultdict(list)
for what_course, value in m.items():
    for rating, ID, TF in value:
        ans[ID].append((what_course, TF))
an = pd.DataFrame(columns=['ID','course_1','course_2'])
for Id, ar in ans.items():
    if len(ar) == 1:
        
        need_new = ar[0][1]
        arr = []
        if need_new:
            arr = [Id, ar[0][0], '???']
        else:
            arr = [Id, ar[0][0], '-']
        df1 = pd.DataFrame(data =arr, columns = ['ID','course_1','course_2'])
        an = an.append(df1, ignore_index=True)
    else:
        arr = [Id, ar[0][0], ar[1][0]]
        df1 = pd.DataFrame(columns = ['ID','course_1','course_2'], data=arr) 
        an = an.append(df1, ignore_index=True)
an.head()
        #    * Осенние: везде 30 мест, кроме Statistical Learning
#Theory (60 мест), Высокопроизводительных вычислений (60 мест), Анализа неструктурированных данных ($\infty$ мест)   


ValueError: Shape of passed values is (3, 1), indices imply (3, 3)

Здесь для проверки приведена таблица, в которой есть 2 дополнительные колонки:
    
    1) is_first_place - является ли студент лучшим по перцентили хотя бы на одном из курсов, куда он был зачислен 
    (True / NaN)
    
    2) is_last_place  - является ли студент худшим по перцентили хотя бы на одном из курсов, куда он был зачислен (True / NaN)

In [ ]:
!wget  -O '2_task_check.csv' -q 'https://www.dropbox.com/s/v8o2zzq3iz5gc9w/_2_task_check.csv?dl=0'
check_df = pd.read_csv('2_task_check.csv')

После распределения студентов в соответствии с первым приоритетом добавьте в свой датафрейм аналогичные признаки и запустите проверку:

In [ ]:
fir_col_name =       # insert name of your new column with is_first_place as str
last_col_name =      # insert name of your new column with is_last_place as str


assert((df[df[fir_col_name].isna() == False][['id']].sort_values('id').reset_index(drop=True)
        ==
        check_df[check_df['is_first_place'].isna() == False][['id']].sort_values('id').reset_index(drop=True)
       ).id.values).all()


assert((df[df[last_col_name].isna() == False][['id']].sort_values('id').reset_index(drop=True)
       == 
       check_df[check_df['is_last_place'].isna() == False][['id']].sort_values('id').reset_index(drop=True)
      ).id.values).all()

#### 3. [3 балла] Проведите все три волны отбора студентов на курсы по выбору

In [ ]:
# (￣^￣)ゞ

**Отправьте свой файл res_fall.csv в контест (https://contest.yandex.ru/contest/40211/problems/A/) и прикрепите/укажите ниже ваш никнейм и ссылку на успешную посылку.**

*Дисклеймер:*

Успешная посылка в контесте является **обязательным** условием получения полного балла за этот (и следующий) пункт. Если добиться успешной посылки не удастся, баллы будут выставляться на усмотрение проверяюшего. 

При этом ОК в контесте не гарантирует полный балл. Оценка всё равно может быть снижена в случае обнаружения неэффективностей или ошибок в коде. Если вы сдадите в AnyTask очевидно неработающий код или ноутбук без кода, но при этом в контест будет сдан корректный файл, то это будет расцениваться как плагиат.

На всякий случай просим вас сдать вместе с ноутбуком файл res_fall.csv в anytask

**Дополнительное задание. [2 бонусных балла] Распределите таким же образом студентов еще и на весенние курсы по выбору.**

Если ваш код был хорошо структурирован, то это не составит проблем. 

Если вы выполнили это задание, сдайте среди прочего файл res_spring.csv в таком же формате, как и res_fall.csv.

In [ ]:
# (￣^'￣)ゞ

**Отправьте свой файл res_spring.csv в контест (https://contest.yandex.ru/contest/40211/problems/B/) и прикрепите/укажите ниже ваш никнейм и ссылку на успешную посылку.**

На всякий случай просим вас сдать вместе с ноутбуком файл res_spring.csv в anytask

Вставьте картинку, описывающую ваш опыт выполнения этого задания: